test/functional/terminal/tui_spec.lua

```diff
@@ -33,6 +33,92 @@ local assert_log = t.assert_log
 
 local testlog = 'Xtest-tui-log'
 
+describe('TUI :detach', function()
+  before_each(function()
+    os.remove(testlog)
+  end)
+  after_each(function()
+    -- os.remove(testlog)
+  end)
+
+  it('does not stop server', function()
+    local server_super = n.clear()
+    local client_super = n.new_session(true)
+    finally(function()
+      server_super:close()
+      client_super:close()
+    end)
+
+    local child_server = new_pipename()
+    local screen = tt.setup_child_nvim({
+      '--listen',
+      child_server,
+      '-u',
+      'NONE',
+      '-i',
+      'NONE',
+      '--cmd',
+      'colorscheme vim',
+      '--cmd',
+      nvim_set .. ' notermguicolors laststatus=2 background=dark',
+    }, {
+      env = {
+        NVIM_LOG_FILE = testlog,
+      },
+    })
+
+    tt.feed_data('iHello, World')
+    screen:expect {
+      grid = [[
+      Hello, World^                                      |
+      {4:~                                                 }|*3
+      {MATCH:No Name}
+      {3:-- INSERT --}                                      |
+      {3:-- TERMINAL --}                                    |
+    ]],
+    }
+
+    local child_session = n.connect(child_server)
+    local status, child_uis = child_session:request('nvim_list_uis')
+    assert(status)
+    eq(1, #child_uis)
+
+    tt.feed_data('\027\027:detach\013')
+    -- Note: "Process exited" message is misleading; tt.setup_child_nvim() sees the foreground
+    -- process (client) exited, and doesn't know the server is still running?
+    screen:expect {
+      any = [[Hello, World.*Process exited]],
+    }
+
+    child_uis --[[@type any[] ]] = ({ child_session:request('nvim_list_uis') })[2]
+    eq(0, #child_uis)
+
+    -- NOTE: The tt.setup_child_nvim() screen just wraps :terminal, it's not connected to the child.
+    -- To use it again, we need to detach the old one.
+    screen:detach()
+
+    -- Edit some text on the headless server.
+    status = (child_session:request('nvim_input', 'ddiWe did it, pooky.<Esc><Esc>'))
+    assert(status)
+
+    -- Test reattach by connecting a new TUI.
+    local screen_reattached = tt.setup_child_nvim({
+      '--remote-ui',
+      '--server',
+      child_server,
+    })
+    screen_reattached:expect {
+      grid = [[
+      We did it, pooky^.                                 |
+      {4:~                                                 }|*3
+      {5:[No Name] [+]                                     }|
+                                                        |
+      {3:-- TERMINAL --}                                    |
+    ]],
+    }
+  end)
+end)
+
 if t.skip(is_os('win')) then
   return
```

Comment:
Not good that we are skipping all these tests on Windows. But at least the new tests above should run there.